geocoder/geocoder.cpp

```diff
@@ -172,39 +205,92 @@ void Geocoder::Go(Context & ctx, Type type) const
 
       subquery.push_back(ctx.GetToken(j));
 
-      auto const * entries = m_hierarchy.GetEntries(subquery);
-      if (!entries || entries->empty())
-        continue;
-
       Layer curLayer;
       curLayer.m_type = type;
-      for (auto const & e : *entries)
-      {
-        if (e.m_type != type)
-          continue;
 
-        if (ctx.GetLayers().empty() || FindParent(ctx.GetLayers(), e))
-          curLayer.m_entries.emplace_back(&e);
+      // Buildings are indexed separately.
+      if (type == Type::Building)
+      {
+        FillBuildingsLayer(ctx, subquery, curLayer);
       }
-
-      if (!curLayer.m_entries.empty())
+      else
       {
-        ScopedMarkTokens mark(ctx, type, i, j + 1);
+        FillRegularLayer(ctx, type, subquery, curLayer);
+      }
+
+      if (curLayer.m_entries.empty())
+        continue;
 
-        double const certainty =
-            static_cast<double>(ctx.GetNumUsedTokens()) / static_cast<double>(ctx.GetNumTokens());
+      ScopedMarkTokens mark(ctx, type, i, j + 1);
 
-        for (auto const * e : curLayer.m_entries)
-          ctx.AddResult(e->m_osmId, certainty);
+      // double const certainty =
+      //     static_cast<double>(ctx.GetNumUsedTokens()) /
+      //     static_cast<double>(ctx.GetNumTokens());
 
-        ctx.GetLayers().emplace_back(move(curLayer));
-        SCOPE_GUARD(pop, [&] { ctx.GetLayers().pop_back(); });
+      double certainty = 0;
+      for (auto const t : ctx.GetTokenTypes())
+      {
+        auto const it = kWeight.find(t);
+        CHECK(it != kWeight.end(), ());
+        certainty += it->second;
```

Comment:
Am I right that certainity is a log prob?

Reply:
No.
certainty is described in geocoder/result.hpp
It is just a number in [0.0, 1.0] at this stage, no more guarantees are given.

Reply:
But why not to transform it to log-prob? It'll have clear semantic.

Reply:
Can you explain what you mean in more detail?
Here is what certainty is now:
1. Let's see how many query tokens are covered with the current parse and assign this fraction to certainty.
2. Turns out, it works badly since a single matched long street in the query (i.e., wrong city, wrong region, wrong locality, correct street) can shadow a more relevant result (correct city, correct locality, wrong street) in the case where we do not have good results in our database anyway. So let's make some parts of the query heavier (heuristically). This turns out to work more predictable.

Reply:
I added this discussion as the documentation to GetWeight in https://github.com/mapsme/omim/pull/10815